ml/nemo_xgboost_optimization.py

```diff
@@ -0,0 +1,337 @@
+# MAIN FUNCTION
+# ########################################################
+argv = sys.argv[1:]
+tablename = ''
+try:
+  opts, args = getopt.getopt(argv,"ht:",["tablename="])
+except getopt.GetoptError:
+  print('nemo_xgboost_optimization.py -t <tablename>')
+  sys.exit(2)
+for opt, arg in opts:
+  if opt == '-h':
+    print('nemo_xgboost_optimization.py -t <tablename>')
+    sys.exit()
+  elif opt in ("-t", "--tablename"):
+    tablename = arg
+
+modelname = tablename + "_bst.model"
+processed_rows = load_data_from_db(tablename)
+# write_to_file('process_test', processed_rows)
+
+## Make Dictionary
+col = []
+for row in processed_rows:
+  arr = row.split()
+  for it in arr[1:]:
+    k,v = it.split(':')
+    col.append(int(k))
+le = preprocessing.LabelEncoder()
+ids = le.fit_transform(col)
+col_to_id = dict(zip(col, ids))
+id_to_col = dict(zip(ids, col))
+
+encoded_rows = encode_processed_rows(processed_rows, col_to_id)
+write_to_file('nemo_optimization.out', encoded_rows)
+# write_to_file('decode_test', decode_rows(encoded_rows, id_to_col))
+ddata = xgb.DMatrix('nemo_optimization.out')
+
+avg_20_duration = np.mean(ddata.get_label()[:20])
+print("average job duration: ", avg_20_duration)
+allowance = avg_20_duration // 25  # 4%
+
+row_size = len(processed_rows)
+print("total_rows: ", row_size)
+
+## TRAIN
+dtrain = ddata.slice([i for i in range(0, row_size) if i%6 != 5])  # mod is not 5
+print("train_rows: ", dtrain.num_row())
+dtest = ddata.slice([i for i in range(0, row_size) if i%6 == 5])  # mod is 5
+print("test_rows: ", dtest.num_row())
+labels = dtest.get_label()
+
+## Existing booster
+bst_opt = xgb.Booster(model_file=modelname) if Path(modelname).is_file() else None
+preds_opt = bst_opt.predict(dtest) if bst_opt is not None else None
+error_opt = (sum(1 for i in range(len(preds_opt)) if abs(preds_opt[i] - labels[i]) > allowance) / float(len(preds_opt))) if preds_opt is not None else 1
+print('opt_error=%f' % error_opt)
+
+learning_rates = [0.1, 0.3, 0.4, 0.5, 0.6, 0.7, 0.9]
+for lr in learning_rates:
+  param = {'max_depth':6, 'eta':lr, 'verbosity':0, 'objective':'reg:linear'}
+
+  watchlist = [(dtest, 'eval'), (dtrain, 'train')]
+  num_round = row_size // 10
+  bst = xgb.train(param, dtrain, num_round, watchlist, early_stopping_rounds=5)
+
+  preds = bst.predict(dtest)
+  error = (sum(1 for i in range(len(preds)) if abs(preds[i] - labels[i]) > allowance) / float(len(preds))) if len(preds) > 0 else 1.0
+  print('error=%f' % error)
+
+  ## Better booster
+  if error <= error_opt:
+    bst_opt = bst
+    bst.save_model(modelname)
+
+## Let's now use bst_opt
+# fscore = bst_opt.get_fscore()
+# sorted_fscore = sorted(fscore.items(), key=lambda kv: kv[1])
+# for i in range(len(sorted_fscore)):
+#   print("\nSplit Value Histogram:")
+#   feature = sorted_fscore.pop()[0]
+#   print(feature, "=", id_to_col[int(feature[1:])])
+#   hg = bst_opt.get_split_value_histogram(feature)
+#   print(hg)
```

Comment:
Many commented lines..
Maybe only you have usages of all methods.. let's improve the script incrementally.

Reply:
This was there for visualization for the users who want to get a look of the model. I'll make it so that it could be used optionally in the future.